### PR TITLE
k8s: check $number_cpus before using it

### LINKED ISSUE #4328: kubernetes/k8s-number-cpus.bats: ensure `$number_cpus` is defined before use
# Description of problem

In the algorithm of `integration/kubernetes/k8s-number-cpus.bats` the current number of cpus is saved in `$number_cpus` then compared with the maximum allowed (`$max_number_cpus`). However the `$number_cpus` value might be undefined at some point in time and this will break the script.   

Example of failure drove by that behavior:
```
not ok 1 Check number of cpus
# (in test file k8s-number-cpus.bats, line 35)
#   `[ "$number_cpus" -le "$max_number_cpus" ]' failed with status 2
# [bats-exec-test:23] INFO: k8s configured to use runtimeclass
# pod/cpu-test created
# pod/cpu-test condition met
# /tmp/bats-run-R6kbge/bats.262021.src: line 35: [: : integer expression expected
# Name:         cpu-test
# Namespace:    default
# Priority:     0
# Node:         fedora32.localdomain/192.168.122.91
# Start Time:   Fri, 07 Jan 2022 13:49:11 +0000
# Labels:       <none>
# Annotations:  <none>
# Status:       Running
# IP:           10.244.0.26
# IPs:
#   IP:  10.244.0.26
# Containers:
#   c1:
#     Container ID:  cri-o://cc4fa93fec804564a5393c2266da26af8db461d948b6a45887845e92b221948e
#     Image:         quay.io/prometheus/busybox:latest
#     Image ID:      quay.io/prometheus/busybox@sha256:2548dd93c438f7cf8b68dc2ff140189d9bcdae7130d3941524becc31573ec9e3
#     Port:          <none>
#     Host Port:     <none>
#     Command:
#       tail
#       -f
#       /dev/null
#     State:          Running
#       Started:      Fri, 07 Jan 2022 13:49:17 +0000
#     Ready:          True
#     Restart Count:  0
#     Limits:
#       cpu:  500m
#     Requests:
#       cpu:        500m
#     Environment:  <none>
#     Mounts:
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-k8zhh (ro)
#   c2:
#     Container ID:  cri-o://a18e4e68400a13c032efaa187b2c56b56696229c6734172abf5db403a21da2ea
#     Image:         quay.io/prometheus/busybox:latest
#     Image ID:      quay.io/prometheus/busybox@sha256:2548dd93c438f7cf8b68dc2ff140189d9bcdae7130d3941524becc31573ec9e3
#     Port:          <none>
#     Host Port:     <none>
#     Command:
#       sleep
#       10
#     State:          Running
#       Started:      Fri, 07 Jan 2022 13:49:20 +0000
#     Ready:          True
#     Restart Count:  0
#     Limits:
#       cpu:  500m
#     Requests:
#       cpu:        500m
#     Environment:  <none>
#     Mounts:
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-k8zhh (ro)
# Conditions:
#   Type              Status
#   Initialized       True 
#   Ready             True 
#   ContainersReady   True 
#   PodScheduled      True 
# Volumes:
#   kube-api-access-k8zhh:
#     Type:                    Projected (a volume that contains injected data from multiple sources)
#     TokenExpirationSeconds:  3607
#     ConfigMapName:           kube-root-ca.crt
#     ConfigMapOptional:       <nil>
#     DownwardAPI:             true
# QoS Class:                   Burstable
# Node-Selectors:              <none>
# Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
#                              node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
# Events:
#   Type    Reason     Age   From               Message
#   ----    ------     ----  ----               -------
#   Normal  Scheduled  12s   default-scheduler  Successfully assigned default/cpu-test to fedora32.localdomain
#   Normal  Pulling    10s   kubelet            Pulling image "quay.io/prometheus/busybox:latest"
#   Normal  Pulled     7s    kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 3.467201901s
#   Normal  Created    7s    kubelet            Created container c1
#   Normal  Started    6s    kubelet            Started container c1
#   Normal  Pulling    6s    kubelet            Pulling image "quay.io/prometheus/busybox:latest"
#   Normal  Pulled     3s    kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 3.246026645s
#   Normal  Created    3s    kubelet            Created container c2
#   Normal  Started    3s    kubelet            Started container c2
# pod "cpu-test" deleted
[run_kubernetes_tests.sh:92] ERROR: bats k8s-number-cpus.bats
```

### PR DESCRIPTION
**Pull request by wainersm**
In integration/kubernetes/k8s-number-cpus.bats the number of cpus in
the pod is stored in $number_cpus which is afterwards used to compare
with the maximum allowed. But $number_cpus can be empty sometimes, so
here is added a checking before using it.

Fixes #4328
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

### CODE REVIEW

File: integration/kubernetes/k8s-number-cpus.bats
Comment (jodh-intel) [inline integration/kubernetes/k8s-number-cpus.bats:34]: What if the system has 10 cpus? I'd extend the regex to allow that.

Comment (wainersm) [reply integration/kubernetes/k8s-number-cpus.bats:34]: True. I will change it to match multiple digits.

**Comment by wainersm** [comment]
Updated this PR so that:
 * the regex will match multiple digits numbers (@jodh-intel 's suggestion)

File: integration/kubernetes/k8s-number-cpus.bats
Comment (wainersm) [reply integration/kubernetes/k8s-number-cpus.bats]: Done

**Review by jodh-intel — APPROVED** [review]
Thanks @wainersm.

lgtm

**Comment by jodh-intel** [comment]
One more ack here and this PR can land...

File: integration/kubernetes/k8s-number-cpus.bats
Comment (amshinde) [inline integration/kubernetes/k8s-number-cpus.bats:34]: @wainersm Why is this empty sometimes?

Comment (wainersm) [reply integration/kubernetes/k8s-number-cpus.bats:34]: Hi @amshinde , that's actually a good question and I don't know. Worth a separate issue?

The test has a pod with c1 and c2 containers. The former is always spinning but the later dies every 10 seconds, thus c2 is restarted by k8s from time to time. My first thought was that maybe by the time it read from /proc/cpuinfo on c1, the runtime is hot-unplugging the CPU from (dead) c2 thus the proc file is still in undetermined state. I really don't know if such as hypothesis makes sense...

@c3d any clue?

Comment (wainersm) [reply integration/kubernetes/k8s-number-cpus.bats:34]: @gkurz do you know what may be happening?

Comment (egernst) [reply integration/kubernetes/k8s-number-cpus.bats:34]: Weird -- why not just sleep infinity in that workload?

Also -- related -- there seems to be a LOT of overlap between this test and k8s-cpu-ns.bats -- perhaps we should consolidate these two tests?

Comment (egernst) [reply integration/kubernetes/k8s-number-cpus.bats:34]: Also, it may be useful to augment the debug information available as part of the teardown. Either printing number of CPUs, or re-running the kubectl exec exec "$pod_name" -c "$container_name" -- sh -c "$num_cpus_cmd" ?

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

**Review by GabyCT — APPROVED** [review]

**Pull request merged**
